File: src/main/go/leetcode/editor/cn/permutations/Permutations.go
//46.permutations
//给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。
//
//
//
// 示例 1：
//
//
//输入：nums = [1,2,3]
//输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
//
//
// 示例 2：
//
//
//输入：nums = [0,1]
//输出：[[0,1],[1,0]]
//
//
// 示例 3：
//
//
//输入：nums = [1]
//输出：[[1]]
//
//
//
//
// 提示：
//
//
// 1 <= nums.length <= 6
// -10 <= nums[i] <= 10
// nums 中的所有整数 互不相同
//
// Related Topics 数组 回溯
// 👍 1440 👎 0

package main

import (
	"fmt"
)

// leetcode submit region begin(Prohibit modification and deletion)
func permute(nums []int) [][]int {
	var ans [][]int
	permute_(nums, len(nums), 0, &ans)
	return ans
}

// 要确定其中的一种排列,就必须知道数组中每个位置的元素
// 首先，确定第一个位置，有n种可能，即依次将其他位置的元素与首位置调换
// 每次确定当前位置的元素后，接着递归确定其后位置上的元素
// 当最后一个位置的元素确定时，便得出其中的一种排列
func permute_(nums []int, l int, k int, ans *[][]int) {
	if k == l {
		*ans = append(*ans, append([]int{}, nums...))
	}
	for i := k; i < l; i++ {
		swap(nums, k, i)
		permute_(nums, l, k+1, ans)
		swap(nums, k, i)
	}
}

// 根据校验数组和原始数组计算排列数组
func swap(nums []int, i, j int) {
	//还原排列数组
	temp := nums[i]
	nums[i] = nums[j]
	nums[j] = temp
}

//leetcode submit region end(Prohibit modification and deletion)

// test from here
func main() {
	fmt.Println(permute([]int{1, 1, 2}))
}

File: src/main/java/leetcode/editor/cn/Permutations.java
//46.给定一个 没有重复 数字的序列，返回其所有可能的全排列。
//
// 示例: 
//
// 输入: [1,2,3]
//输出:
//[
//  [1,2,3],
//  [1,3,2],
//  [2,1,3],
//  [2,3,1],
//  [3,1,2],
//  [3,2,1]
//] 
// Related Topics 回溯算法 
// 👍 1204 👎 0


package leetcode.editor.cn;

import java.util.ArrayList;
import java.util.List;

public class Permutations {
    public static void main(String[] args) {
        Solution solution = new Permutations().new Solution();
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public List<List<Integer>> permute(int[] nums) {
            List<List<Integer>> list = new ArrayList<>();
            permute(nums, 0, list);
            return list;
        }


        /**
         * 对字符数组进行全排列，递归确定位置k上的元素
         *
         * @param arr
         * @param k
         */
        private int permute(int[] arr, int k, List<List<Integer>> result) {
            if (k == arr.length - 1) {
                // 此时，数组最后一个位置k上的元素已确定，从而确定了一种排列
                result.add(toList(arr));
                return 1;
            }
            int count = 0;
            // 位置k上的元素有ch.Length-k+1种可能值
            // 即k和其后位置上的元素都可能放在k位置上
            for (int i = k; i < arr.length; i++) {
                // 确定i位置上的字符是不是第一次出现，
                // 如果不是第一次出现，则相同的字符已经调换过
                if (this.isFirstChar(arr, k, i)) {
                    this.swap(arr, k, i);
                    count += this.permute(arr, k + 1, result);
                    this.swap(arr, k, i);
                }
            }
            return count;
        }

        /**
         * 交换字符数组元素
         *
         * @param arr
         * @param i
         * @param j
         */
        private void swap(int[] arr, int i, int j) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }

        /**
         * 查找从位置begin开始，end位置上的元素是不是第一次出现
         *
         * @param arr
         * @param begin
         * @param end
         * @return
         */
        private boolean isFirstChar(int[] arr, int begin, int end) {
            if (end > begin) {
                for (int i = begin; i < end; i++) {
                    if (arr[i] == arr[end]){
                        return false;
                    }
                }
            }
            return true;
        }

        private List<Integer> toList(int[] nums) {
            List<Integer> list = new ArrayList<>(nums.length);
            for (int num : nums) {
                list.add(num);
            }
            return list;
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}